### ISSUE TITLE
Carouel max translation 

### ISSUE THREAD

**Issue by nhall** [Formstone/Formstone#253]
This might be difficult to type but let me give it a try. I have an idea for an enhancement and I wanted to get your take on how you might solve it.

The carousel currently has this really nice behavior where it will only translate the canister enough to show the last item within the viewport. Here's the example, I have three items and I have carousel set to show two items at a time. When I hit next it will only translate the canister so that the right edge of the third item is within viewport (as opposed to the left edge of the item reaching the left edge of the viewport). 

I think this is ideal and how it should behave. However, this is dependent on showing two full items in the viewport. Let's say I have the same carousel on a small screen and I want to show all of the first item and have the rest bleed out of the screen. My first instinct is to change the viewport to the size I want the items to be and show one item at a time. Since I am showing one at time when I advance to the last item it will move it within the viewport potentially leaving a bunch of white space after the carousel. This is expected since I set the viewports width. What I've tried to do to get around this is to leave the viewport untouched (so it's auto) and then change the max-width of the item to the maximum size I want with CSS. When I do this advancing to the last item will still move it all way so the left edge of the item meets the left edge of the viewport. Instead of the right edge of the item meeting the right edge of the viewport.

I can create a demo if this unclear. Let me know if you have any ideas on where to start.

Thanks!

**Comment by benplum** [comment]
I think I understand what you are describing, but an example would certainly help illustrate what you're trying to achieve. 

Some background: The positioning is always based on the [left edge](https://github.com/Formstone/Formstone/blob/master/src/js/carousel.js#L482) of the first item in the current page (or right edge, if RTL). Setting a max-width in CSS probably won't change the total movement of the canister based on the current positioning logic. 

My initial thought is that the `matchWidth: false` option could be updated to function in this way (right now it's more like the `paged: true` where the canister snaps to each item instead of pages/groups of items that would fit). I'll look into how we can calculate the pages based on how many elements fit into the viewport based on the natural width set via CSS. This should fix the negative space caused by aligning to the left edge of the last page.

**Comment by nhall** [comment]
I'll get an example together. Thanks!

**Comment by nhall** [comment]
I put together a VERY basic [example](https://codepen.io/nhall/pen/ZRebOg).

Take a look at the first carousel at a screen size larger than 740px. The viewport is set to show three items. If you advance it once it moves the carousel the number of visible items (three) so the fourth item is in the leftmost of the viewport. However, on advancement again it moves the carousel only two items instead of three so the viewport continues to show three items. This is great. I love that I'm not left with empty horizontal space.

Resize you browser down below 740px and you can see the two carousels set to show one item. In the top carousel I've set the viewport width to control the width of the items. When you advance to the end it moved the last item into the viewport leaving empty space to the right. Totally expected and the correct behavior. In order to try and get rid of that space I've left the viewport width untouched in the second example. To control the width of the items I've set a max-width on the items. Since the carousel is set to show one at a time it still wants to move the last slide all the way to left where I'd love to only move it to the right and not have that space.

Does that make sense?

**Comment by benplum** [comment]
Yup! Thanks for the example, this is what I thought you meant but the visualization helps. 

I do think this is related to how 'pages' are aligned based on the left (or right in RTL) edge. I think a new method for the page groupings is in order.

**Comment by benplum** [comment]
I spent some time reworking how the carousel pages are organized. Here's an updated pen: https://codepen.io/benplum/pen/gKmJXM

You will have to set `matchWidth: false` to less CSS determine the page widths. Let me know how this works and we can get a fresh version tagged.

**Comment by nhall** [comment]
I think this works great! It's exactly what I was describing. Thanks!

**Comment by benplum** [comment]
This is up on v 1.4.8.

**Comment by matthisamoto** [comment]
I think something to do with this broke how it calculates what's visible? I have a carousel that's hiding/showing content based on fs-carousel-visible and it's only showing the final two, aligned right but hiding the others and it makes it sorta funny lookin.

[Here's an example](#), the thumbnail area if you hit the arrow till you get to the end.

**Comment by matthisamoto** [comment]
@benplum just want to circle back around to this I don't know if it ever notified you of this since it's closed?

**Comment by nhall** [comment]
Hey Matt,
I took a look at the link. Is it resolved because you are using an older version of Formstone now?

I've been looking through the changes and I'm not totally sure where the issue is. However, I am seeing a problem with the way it's getting left position? Looking at the example that Ben posted, https://codepen.io/benplum/pen/gKmJXM. If I resize my screen below the min-width of the items (320px) I will get a JS error:

`Uncaught TypeError: Cannot read property 'left' of undefined`

Which looks like it's being calculated here:
https://github.com/Formstone/Formstone/blob/master/src/js/carousel.js#L457

Still looking into this.

**Comment by matthisamoto** [comment]
Nick,

I had to change how it was working in order to get the project launched on time. I was using fs-carousel-visible to change the opacity of the items, but I backed off of that because it would only highlight the last "page" of results, meaning if I was showing 3 items, and had a remainder of 1, it would show that remainder at the right edge but the other items from the previous "page" would be invisible so it would be like a single item floating in the distance. So the example is out of date because of this. I'll have to throw together a codepen about it or something when I have a minute.
